src/main/java/com/google/cloud/tools/DeployAction.java

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.tools;
+
+import com.google.cloud.tools.ProcessCaller.Tool;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Deploys an app to the server.
+ */
+public class DeployAction extends Action {
+
+  private static Option[] acceptedFlags = {
+      Option.BUCKET,
+      Option.DOCKER_BUILD,
+      Option.FORCE,
+      Option.IMAGE_URL,
+      Option.PROMOTE,
+      Option.SERVER
+  };
+  private AppType appType;
+  private Action stageAction;
```

Comment:
With the current model, you can stage and not deploy. But you can't deploy without staging, because staging is a dependency for deployment.

If staging an application is not required sometimes, the library should be able to detect that and stage/not stage an app. I think it doesn't make sense to leave this complexity to the user, it should live in the library.

Regarding error messaging, wouldn't it be the case that any failures at the staging level would still show in the stack trace inside StageAction? We can also make it more obvious in the exception message that the issue happened at the staging level. Is there anything I'm missing?